great_expectations/datasource/fluent/invalid_datasource.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING, ClassVar, List, NoReturn, Type, overload
+
+from great_expectations.compatibility import pydantic
+from great_expectations.compatibility.pydantic import Field
+from great_expectations.compatibility.typing_extensions import override
+from great_expectations.datasource.fluent import (
+    DataAsset,
+    Datasource,
+    GxDatasourceWarning,
+    TestConnectionError,
+)
+from great_expectations.datasource.fluent.type_lookup import TypeLookup, ValidTypes
+
+if TYPE_CHECKING:
+    from great_expectations.core.batch_config import BatchConfig
+    from great_expectations.datasource.fluent.batch_request import BatchRequest
+
+
+class GxInvalidDatasourceWarning(GxDatasourceWarning):
+    """
+    A warning that the Datasource configuration is invalid and will must be updated before it can used.
+    """
+
+
+class InvalidAsset(DataAsset):
+    """
+    A DataAsset that is invalid.
+    The DataAsset itself may be valid, but it is classified as invalid because its parent Datasource or sibling assets are invalid.
+    """
+
+    type: str = "invalid"
+    name: str = "invalid"
+
+    class Config:
+        extra = "allow"
+        arbitrary_types_allowed = True
+
+
+class InvalidAssetTypeLookup(TypeLookup):
+    """A TypeLookup that always returns InvalidAsset for any type."""
+
+    @overload
+    def __getitem__(self, key: str) -> Type:
+        ...
+
+    @overload
+    def __getitem__(self, key: Type) -> str:
+        ...
+
+    @override
+    def __getitem__(self, key: ValidTypes) -> ValidTypes:
```

Comment:
Why don't we always return an `InvalidAsset`?

Reply:
Was trying to satisfy the interface of `TypeLookup` to return an actual `Type`/`class` if a string was provided and the classes `type: str` value if a type was provided.
But in this case, we should never actually need to lookup the `invalid` str for `InvalidType` so I can just raise a `NotImplementedError` in that case because it almost certainly is being used incorrectly.